R/ivreg_marg.R

```diff
@@ -0,0 +1,124 @@
+#' Predicted Margins for `ivreg` objects from the \code{AER} package
+#'
+#' Obtains predicted margins and standard errors
+#' of those predictions from a fitted \code{ivreg} model object.
+#'
+#' @inheritParams marg
+#' @param ... additional parameters passed to \code{?marg}.
+#'
+#' @export
+#' @examples
+#' data(margex)
+#' margex$assign <- margex$treatment
+#'
+#' # Probability of getting treatment increases with age
+#' margex$pr_treat <- plogis(
+#'  scale(margex$age) + scale(margex$distance))
+#'
+#' # One-way non-compliance
+#' margex$actual <- margex$assign
+#' margex$actual[margex$assign == 1] <- rbinom(
+#'   n = sum(margex$assign == 1), size = 1,
+#'   prob = margex$pr_treat[margex$assign == 1]
+#' )
+#'
+#' table(margex$assign, margex$actual)
+#'
+#' # Fit the model
+#' mod <- AER::ivreg(
+#'   y ~ as.factor(actual) + age + distance |
+#'     as.factor(assign) + age + distance,
+#'   data = margex)
+#'
+#' # Get margins for different levels of "actual" treatment
+#' modmarg::marg(mod, data = margex, var_interest = 'actual')
+#'
+#' # TODO: weights, clustering
+#'
+marg.ivreg <- function(mod, var_interest,
+                       data,
+                       weights = mod$weights,
+                       ...){
+
+  .marg(mod = mod, var_interest = var_interest,
+        data = data,
+        weights = weights,
+        ...)
+}
+
+pred_se.ivreg <- function(model,
+                          deriv_func = function(x) rep.int(1, length(x)),
+                          link_func = identity,
+                          ...){
+
+  .pred_se(model = model,
+           deriv_func = function(x) rep.int(1, length(x)),
+           link_func = identity,
+           ...)
+
+}
+
+#' @importFrom stats get_all_vars
+get_data.ivreg <- function(model, data, weights){
+
+  # Store original number of rows
+  nrow_orig <- nrow(data)
+
+  # Grab only necessary variables
+  data <- data[, all.vars(model$formula)]
+
+  # Add weights
```

Comment:
Everything from here on below seems similar to `glm`. Can we create a helper function and change both of those?